taichi/analysis/same_statements.cpp

```diff
@@ -0,0 +1,234 @@
+  }
+
+#define DEFINE_FIELD_CHECK(field)    \
+  if (stmt->field != other->field) { \
+    same = false;                    \
+    return;                          \
+  }
+
+#define DEFINE_MAP_ID map_id(stmt->id, other->id);
+
+#define DEFINE_BASIC_CHECK(Type) \
+  DEFINE_TYPE_CHECK(Type)        \
+  DEFINE_OPERAND_CHECK           \
+  DEFINE_MAP_ID
+
+// Compare if two IRNodes are equivalent.
+class IRNodeComparator : public IRVisitor {
+ private:
+  IRNode *other_node;
+  std::map<int, int> id_map; // map the id from this node to the other node
+  std::set<int> captured_id; // ids which don't belong to either node
+
+ public:
+  bool same;
+
+  explicit IRNodeComparator(IRNode *other_node) : other_node(other_node) {
+    same = true;
+  }
+
+  void map_id(int this_id, int other_id) {
+    if (captured_id.find(this_id) != captured_id.end()
+        || captured_id.find(other_id) != captured_id.end()) {
+      same = false;
+      return;
+    }
+    auto it = id_map.find(this_id);
+    if (it == id_map.end()) {
+      id_map[this_id] = other_id;
+    } else if (it->second != other_id) {
+      same = false;
+    }
+  }
+
+  int get_other_id(int this_id) {
+    // get the corresponding id in the other node
+    auto it = id_map.find(this_id);
+    if (it != id_map.end()) {
+      return it->second;
+    }
+    // if not found, should be captured
+    // (What if this_id belongs to the other node? Ignoring this case here.)
+    if (captured_id.find(this_id) == captured_id.end()) {
+      captured_id.insert(this_id);
+    }
+    return this_id;
+  }
+
+  void visit(Block *stmt_list) override {
+    DEFINE_TYPE_CHECK(Block)
+    if (stmt_list->size() != other->size()) {
+      same = false;
+      return;
+    }
+    for (int i = 0; i < (int)stmt_list->size(); i++) {
+      other_node = other->statements[i].get();
+      stmt_list->statements[i]->accept(this);
+      if (!same)
+        break;
+    }
+    other_node = other;
+  }
+
+  void visit(AssertStmt *stmt) override {
+    DEFINE_BASIC_CHECK(AssertStmt)
+    DEFINE_FIELD_CHECK(text)
+  }
+
+  void visit(SNodeOpStmt *stmt) override {
+    DEFINE_BASIC_CHECK(SNodeOpStmt)
+    DEFINE_FIELD_CHECK(type_hint())
+    DEFINE_FIELD_CHECK(op_type)
+    DEFINE_FIELD_CHECK(snode->id)
+  }
+
+  void visit(AllocaStmt *stmt) override {
+    DEFINE_BASIC_CHECK(AllocaStmt)
+    DEFINE_FIELD_CHECK(type_hint())
+  }
+
+  void visit(RandStmt *stmt) override {
+    DEFINE_BASIC_CHECK(RandStmt)
+    DEFINE_FIELD_CHECK(type_hint())
+  }
+
+  void visit(UnaryOpStmt *stmt) override {
+    DEFINE_BASIC_CHECK(UnaryOpStmt)
+    DEFINE_FIELD_CHECK(type_hint())
+    DEFINE_FIELD_CHECK(op_type)
+  }
+
+  void visit(BinaryOpStmt *stmt) override {
+    DEFINE_BASIC_CHECK(BinaryOpStmt)
+    DEFINE_FIELD_CHECK(type_hint())
+    DEFINE_FIELD_CHECK(op_type)
+  }
+
+  void visit(TernaryOpStmt *stmt) override {
+    DEFINE_BASIC_CHECK(TernaryOpStmt)
+    DEFINE_FIELD_CHECK(type_hint())
+    DEFINE_FIELD_CHECK(op_type)
```

Comment:
This design duplicates the definition of `Stmt` members. It will be very easy for someone to add a field to a stmt and then forget to modify here, leaving two parts async... Maybe there are better solutions.

Reply:
But it's not easy to detect the fields automatically. Some fields shouldn't be checked, and some fields should be checked specifically (`snode->id` for example, if I understand this part of the code framework correctly)

Reply:
How about
```cpp
#define DEF_VISITOR(XStmt) \
void visit(XStmt *stmt) override { \
    DEFINE_BASIC_CHECK(XStmt) \
    DEFINE_FIELD_CHECK(type_hint()) \
    DEFINE_FIELD_CHECK(op_type)
```

Reply:
Good idea, but I think it looks clearer to explicitly expose the function signature `void visit(XStmt *stmt)` and the fields `type_hint()` and `op_type` so that developers know what it does by only looking at the function.

Maybe `DEFINE_FIELD_CHECK(type_hint(), op_type)` is an option to simplify the code, but I don't know how to write such a macro :)

Reply:
See https://github.com/yuanming-hu/taichi/blob/master/taichi/common/serialization.h and https://github.com/yuanming-hu/taichi_mpm/blob/master/src/mpm.h#L134